### PR TITLE
Register HotRestartingParent as udp forwarding packet recipient

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: Register HotRestartingParent as udp forwarding packet recipient
Additional Description: When listeners start draining during hot restart, pass them a reference to HotRestartingParent so that ActiveQuicListeners can forward non-routable UDP packets to the child instance.
Risk Level: Low, no behaviors are actually changed yet.
Testing: Existing coverage should cover all touched paths (except the TODO one).
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by danzh2010** [comment]
This is still a draft? Do you mean to actually send it out for review?

**Comment by ravenblackx** [comment]
> This is still a draft? Do you mean to actually send it out for review?

I did.

File: source/server/hot_restarting_parent.cc [hunk 1]
```diff
@@ -184,7 +195,11 @@ void HotRestartingParent::Internal::recordDynamics(HotRestartMessage::Reply::Sta
   }
 }
 
-void HotRestartingParent::Internal::drainListeners() { server_->drainListeners(); }
+void HotRestartingParent::Internal::drainListeners() {
+  Network::ExtraShutdownListenerOptions options;
+  options.non_dispatched_udp_packet_handler_ = *this;
```
Comment (danzh2010) [inline source/server/hot_restarting_parent.cc:200]: Is there a hot restarting test verifies that non_dispatched_udp_packet_handler_ is not nullopt?

File: source/server/server.cc [hunk 2]
```diff
@@ -184,9 +184,11 @@ const Upstream::ClusterManager& InstanceImpl::clusterManager() const {
   return *config_.clusterManager();
 }
 
-void InstanceImpl::drainListeners() {
+void InstanceImpl::drainListeners(OptRef<const Network::ExtraShutdownListenerOptions> options) {
   ENVOY_LOG(info, "closing and draining listeners");
-  listener_manager_->stopListeners(ListenerManager::StopListenersType::All);
+  Network::ExtraShutdownListenerOptions empty_options;
+  listener_manager_->stopListeners(ListenerManager::StopListenersType::All,
+                                   options.has_value() ? *options : empty_options);
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/server/server.cc:191]: `initializer_list cannot be used on right hand side of operator :`

[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/server/server.cc]: Worked with `Network::ExtraShutdownListenerOptions{}` though, and in all the other locations {}. I thought it didn't like it when you pass inline-constructed things as references, I guess maybe I'm mixing that up with some test context.

File: source/server/hot_restarting_parent.cc (hunk 1 above)
Comment (ravenblackx) [reply source/server/hot_restarting_parent.cc:200]: Updated a test to do this.

File: test/extensions/listener_managers/listener_manager/listener_manager_impl_test.cc [hunk 3]
```diff
@@ -3339,12 +3338,12 @@ name: foo
...
   EXPECT_CALL(*listener_foo, onDestroy());
-  Network::ExtraShutdownListenerOptions empty_options;
-  manager_->stopListeners(ListenerManager::StopListenersType::All, empty_options);
-  EXPECT_EQ(1, server_.stats_store_.counterFromString("listener_manager.listener_stopped").value());
+  manager_->stopListeners(ListenerManager::StopListenersType::All, {);
+    EXPECT_EQ(1,
+              server_.stats_store_.counterFromString("listener_manager.listener_stopped").value());
 
-  // Validate that adding a listener is not allowed after all listeners are stopped.
-  const std::string listener_bar_yaml = R"EOF(
```
Comment (danzh2010) [inline test/extensions/listener_managers/listener_manager/listener_manager_impl_test.cc:3343]: Something wrong with the formatting in this file?

Comment (ravenblackx) [reply test/extensions/listener_managers/listener_manager/listener_manager_impl_test.cc]: Doh, typo.

**Review by RyanTheOptimist — APPROVED** [review]
Wow, this is incredibly clean! There's a lot of plumbing here, but it seems mostly mechanical. Thanks for splitting this work into easily digestible pieces!

File: envoy/server/listener_manager.h [hunk 4]
```diff
@@ -222,7 +222,8 @@ class ListenerManager {
    * listeners on workers. Once a listener is stopped, any listener modifications are not allowed.
    * @param stop_listeners_type indicates listeners to stop.
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply envoy/server/listener_manager.h:223]: Done.

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**